### PR TITLE
Fix precision issues in CPU remainder

### LINKED ISSUE #24861: Float overflow in torch.remainder
## 🐛 Bug

`torch.remainder` gives the wrong output for very large float dividends on CPU, could be the `float` version of #5875

## To Reproduce
```python
import torch

x = torch.tensor(2749682432.0)
q = 36
print(torch.remainder(x,q))
```

actual output is `128.0` whereas the correct output should be `20`

## More Information

- `x % q` produces the same incorrect output
- `torch.fmod(x,q)` produces the correct output (`20`) 
- constructing the dividend as an integer (`torch.tensor(2749682432.0, dtype=torch.int)`, overflows the dividend making it `-1545284864` and produces the correct output for that integer

This seems like the issue is likely overflow

## Environment
PyTorch version: 0.4.1  
                                                                                                                      Is debug build: No   
                                                                                                                         CUDA used to build PyTorch: None       
                                                                                                                                                                                                                                                                                                                                                                                       
OS: Fedora release 30 (Thirty)                 
GCC version: (GCC) 9.1.1 20190503 (Red Hat 9.1.1-1)                                
CMake version: Could not collect                                                                                                              

Python version: 3.7           
Is CUDA available: No    
CUDA runtime version: No CUDA                     
GPU models and configuration: No CUDA       
Nvidia driver version: No CUDA                 
cuDNN version: No CUDA         
                                                                                                                                                                                                                                                                                                                                                                                                     
Versions of relevant libraries: 
[pip] Could not collect     
[conda] pytorch-cpu               0.4.1                py37_cpu_1    pytorch   
[conda] torchvision-cpu           0.2.1                    py37_1    pytorch

## Additional context

I might be able to make the fix myself, basing off of #5906 but does it still make sense to make the fix in TH if it will be ported to Aten soon anyways #24507 ? Or should I make the port + fix together?

### LINKED ISSUE #37743: torch.remainder gives a remainder larger than the divisor
## 🐛 Bug

`torch.remainder` is wrong on some inputs. It's accurate when computing in float64, but the remainder can be larger than the divisor in float32. What's worse is that the wrong result is **different** on CPU vs GPU

## To Reproduce

```python
import torch
import math
import numpy as np

n = 1e9

torch.remainder(torch.tensor([n], device="cpu", dtype=torch.float32), math.pi)  # will return tensor([-64.]) which is larger than the divisor math.pi

torch.remainder(torch.tensor([n], device="cuda", dtype=torch.float32), math.pi)  # will return tensor([-33.5333], device='cuda:0') which is also larger than the divisor math.pi
```

## Expected behavior
```python
np.remainder(np.array([n], dtype=np.float32), math.pi)  # will return array([1.024195], dtype=float32)
```

Some other observations:
- The results of `torch.fmod` matches the results of `np.fmod` in both GPU and CPU in fp32 and fp64
- Since by default `x % y` will call `torch.remainder(x, y)` if `x` and `y` are PyTorch Tensors, wrong behavior in `torch.remainder` also results in wrong behavior of `%`.

## Environment

```
PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: EndeavourOS Linux
GCC version: (GCC) 8.4.0
CMake version: version 3.17.0

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 440.82
cuDNN version: /usr/lib/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] pytorch-lightning==0.7.5
[pip] torch==1.5.0
[pip] torchcontrib==0.0.2
[pip] torchdiffeq==0.0.1
[pip] torchvision==0.6.0a0+82fd1c8
[conda] blas                      2.15                        mkl    conda-forge
[conda] cudatoolkit               10.2.89              hfd86e86_0  
[conda] libblas                   3.8.0                    15_mkl    conda-forge
[conda] libcblas                  3.8.0                    15_mkl    conda-forge
[conda] liblapack                 3.8.0                    15_mkl    conda-forge
[conda] liblapacke                3.8.0                    15_mkl    conda-forge
[conda] mkl                       2020.0                      166    conda-forge
[conda] numpy                     1.18.1           py37h8960a57_1    conda-forge
[conda] pytorch                   1.5.0           py3.7_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] pytorch-lightning         0.7.5                     dev_0    <develop>
[conda] torchcontrib              0.0.2                    pypi_0    pypi
[conda] torchdiffeq               0.0.1                    pypi_0    pypi
[conda] torchvision               0.6.0                py37_cu102    pytorch
```

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by peterbell10**
Together with #37758, this fixes #37743 and fixes #24861.

This follows the CUDA fix in #37758, vectorised using a `blendv` to replace the if conditionals.

Most of the complication is from `remainder` supporting `at::Half` where `fmod` doesn't. I've now got `fmod` working on `Vec256<at::Half>` as well as enabling half dispatch for `fmod` so it matches `remainder`.

I also added `fmod` support to `Vec256<at::BFloat16>` before realising that `remainder` doesn't support `BFloat16` anyway. I could also enable `BFloat16` if that's desirable. If not, I don't think `Vec256<BFloat16>` should be missing `fmod` anyway.

### CODE REVIEW

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 1]
```diff
@@ -45,6 +45,13 @@ namespace at {
 namespace vec256 {
 // See Note [Acceptable use of anonymous namespace in header]
 namespace {
+// at::Half should be treated as floating point
+template <typename T>
+struct is_floating_point:
+    std::integral_constant<bool,
+      std::is_floating_point<T>::value ||
+      std::is_same<T, at::Half>::value> {
+};
```
Comment (ezyang) [inline aten/src/ATen/cpu/vec256/vec256_base.h:54]: Hmmmm... are you sure we don't already have a concept for this elsewhere in the codebase?

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 2]
```diff
@@ -210,20 +217,20 @@ struct Vec256 {
...
   }
   template <typename float_t_abs = T,
-            typename std::enable_if<std::is_floating_point<float_t_abs>::value, int>::type = 0>
+            typename std::enable_if<is_floating_point<float_t_abs>::value, int>::type = 0>
   Vec256<T> abs() const {
     // float_t_abs is for SFINAE and clarity. Make sure it is not changed.
     static_assert(std::is_same<float_t_abs, T>::value, "float_t_abs must be T");
     // Specifically deal with floating-point because the generic code above won't handle -0.0 (which should result in
     // 0.0) properly.
-    return map(std::abs);
```
Comment (ezyang) [inline aten/src/ATen/cpu/vec256/vec256_base.h:226]: huh, what's going on here?

**Comment by ezyang** [comment]
It would be great if we could have a little before and after perf comparison. The new implementation is going to be slower but it would be good to have a little napkin saying how much slower.

**Comment by ezyang** [comment]
@xwang233 do you think you could help review this?

**Comment by ezyang** [comment]
cc @andreaskoepf perhaps, too

File: aten/src/ATen/cpu/vec256/vec256_base.h (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/cpu/vec256/vec256_base.h:54]: Not as far as I could see. grepping for `is_floating_point` didn't show anything and I had a look in `Half.h`

File: aten/src/ATen/cpu/vec256/vec256_base.h (hunk 2 above)
Comment (peterbell10) [reply aten/src/ATen/cpu/vec256/vec256_base.h:226]: There is no `std::abs(at::Half)` so this allows implicit conversions to float.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 3]
```diff
@@ -115,11 +115,16 @@ void remainder_kernel(TensorIterator& iter) {
...
         [=](scalar_t a, scalar_t b) __ubsan_ignore_float_divide_by_zero__ -> scalar_t {
-          return a - b * at::native::floor_impl(a / b);
+          scalar_t mod = std::fmod(a, b);
+          if ((mod != 0) && ((b < 0) != (mod < 0))) mod += b;
+          return mod;
         },
         [=](Vec256<scalar_t> a, Vec256<scalar_t> b) {
-          Vec256<scalar_t> r = a - b * (a / b).floor();
-          return r;
+          // TODO: Sleef 3.5 will have a Sleef_remainder{d4,f8}
```
Comment (andreaskoepf) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:123]: If `Sleep_remainder()` will act like IEEE 754-style remainder (e.g. like python [math.remainder()](https://docs.python.org/3/library/math.html#math.remainder) or C++ [std::remainder()](https://en.cppreference.com/w/cpp/numeric/math/remainder)) this will be not behave like the '%' operator or `torch.remainder()`, e.g. 
```
>>> math.remainder(3, 4)
-1.0
>>> 3 % 4
3
```
@peterbell10 I do not know if you had some tricky use of the IEEE 754-style remainder in mind? .. if not I would recommend to remove that reference to `Sleef_remainder`.

Comment (peterbell10) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Yes, this was a misunderstanding. I've removed the comment.

**Comment by peterbell10** [comment]
My benchmark shows the new version is ~50-60% slower for both `float` and `double` tensors.

**Review by xwang233 — APPROVED** [review]
Thanks for the PR! Since the remainder tests are passed, this LGTM. I like the use of XOR instead of != in the mask.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 4]
```diff
@@ -544,7 +548,7 @@ void fmod_kernel(TensorIterator& iter) {
       });
     });
   } else {
-    AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "fmod_cpu", [&]() {
+    AT_DISPATCH_FLOATING_TYPES_AND(kHalf, iter.dtype(), "fmod_cpu", [&]() {
```
Comment (andreaskoepf) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:551]: Was there any reason not to use the `AT_DISPATCH_FLOATING_TYPES_AND_HALF` macro, see also ln 573?

File: aten/src/ATen/cpu/vec256/vec256_bfloat16.h [hunk 5]
```diff
@@ -252,6 +252,15 @@ template <> class Vec256<BFloat16> {
   Vec256<BFloat16> expm1() const {
     return map(Sleef_expm1f8_u10);
   }
+  Vec256<BFloat16> fmod(const Vec256<BFloat16> & q) const {
```
Comment (andreaskoepf) [inline aten/src/ATen/cpu/vec256/vec256_bfloat16.h:255]: Is the BFloat16 fmod specialization used somewhere? I saw only the dispatch code in BinaryOpsKernel.cpp for `kHalf` but not for `kBFloat16`?

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 4 above)
Comment (andreaskoepf) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:551]: Maybe should become `AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, iter.dtype(), ... `?

Comment (peterbell10) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:551]: I noticed the deprecation message for `AT_DISPATCH_ALL_TYPES_AND_HALF` and figured the `_AND` variants were preferred.
https://github.com/pytorch/pytorch/blob/87e97ced2004e2ea0e97ed380a020c2281e76204/aten/src/ATen/Dispatch.h#L50-L51

File: aten/src/ATen/cpu/vec256/vec256_bfloat16.h (hunk 5 above)
Comment (peterbell10) [reply aten/src/ATen/cpu/vec256/vec256_bfloat16.h:255]: It's not currently used, but since it's a floating point type it should have that operation available. I can also enable `BFloat16` dispatch if that's desirable.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 4 above)
Comment (andreaskoepf) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:551]: @peterbell10 sry, I was not aware of the deprecation.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 6]
```diff
@@ -115,11 +115,15 @@ void remainder_kernel(TensorIterator& iter) {
...
-          return a - b * at::native::floor_impl(a / b);
+          scalar_t mod = std::fmod(a, b);
+          if ((mod != 0) && ((b < 0) != (mod < 0))) mod += b;
+          return mod;
         },
         [=](Vec256<scalar_t> a, Vec256<scalar_t> b) {
-          Vec256<scalar_t> r = a - b * (a / b).floor();
-          return r;
+          auto mod = a.fmod(b);
+          const auto zero = Vec256<scalar_t>(0);
```
Comment (andreaskoepf) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:124]: I noticed that other functions in BinaryOpsKernel.cpp declare constants outside the lambda, e.g. before the `cpu_kernel_vec()` call, but I doubt this has a perf impact...

**Review by andreaskoepf — APPROVED** [review]

**Comment by ezyang** [comment]
The perf loss is regrettable, but correctness first!
